### ISSUE TITLE
Rule Request: Disallow native bigint operations on polyfilled bigint

### ISSUE THREAD

**Issue by AnyhowStep** [typescript-eslint/typescript-eslint#2021]
I love that `bigint` is part of the ES spec. However, some environments still do not natively support `bigint`. I have projects that need to use `bigint`, and also polyfill it for environments that do not support it (like Safari).

So, the following code would be bad,
```ts
declare const a : bigint;
declare const b : bigint;
console.log(a + b);
```

Because `a, b` are objects if `bigint` is polyfilled. All code manipulating `bigint` should not use native `bigint` operations. They should use wrapper functions, instead.

-----

I **already** have a rule implemented that I use for these projects, but it is **very ugly** and **hacked together**. (Copy pasted stuff and tweaked it)

I would love it if it was polished and made an official part of the typescript-eslint project.

Code for the rule,
```ts
import {
    TSESTree,
    AST_NODE_TYPES,
} from "@typescript-eslint/experimental-utils";
import * as ts from "typescript";
import * as tsutils from "tsutils";
const util = require("@typescript-eslint/eslint-plugin/dist/util");
/*
type Options = [
    {
        //No options for now
    },
];
*/
type MessageId =
    | "errorBuiltInOperatorUsedWithBigInt";

module.exports = util.createRule/*<Options, MessageId>*/({
    name: "polyfilled-bigint",
    meta: {
        type: "suggestion",
        docs: {
            description: "Restricts builtin operators from being used on bigints; as bigints are assumed to be polyfilled",
            category: "Best Practices",
            recommended: false,
            requiresTypeChecking: true,
        },
        schema: [
            {
                type: "object",
                properties: {
                    //No options for now
                },
                additionalProperties: false,
            },
        ],
        messages: {
            errorBuiltInOperatorUsedWithBigInt : (
                "Unexpected bigint used as operand of builtin operator. BigInt assumed to be polyfilled."
            ),
        },
    },
    defaultOptions: [
        {
        },
    ],
    create(context : any) {
        const service = util.getParserServices(context);
        const checker = service.program.getTypeChecker();

        const checkedNodes = new Set<TSESTree.Node>();

        return {
            BinaryExpression : checkBinaryNode,
            UpdateExpression : checkUpdateNode,
            UnaryExpression: checkUnaryNode,
        };

        function checkUnaryNode(node: TSESTree.UnaryExpression): void {
            checkBinaryNode(node.argument, true);
        }

        function checkUpdateNode(node: TSESTree.UpdateExpression): void {
            checkBinaryNode(node.argument, true);
        }

        /**
         * This function analyzes the type of a boolean expression node and checks if it is allowed.
         * It can recurse when checking nested logical operators, so that only the outermost expressions are reported.
         * @param node The AST node to check.
         * @param isRoot Whether it is the root of a logical expression and there was no recursion yet.
         * @returns `true` if there was an error reported.
         */
        function checkBinaryNode(node: TSESTree.Node, isRoot = false): boolean {
            // prevent checking the same node multiple times
            if (checkedNodes.has(node)) {
                return false;
            }
            checkedNodes.add(node);

            // for logical operator, we also check its operands
            if (
                (
                    node.type === AST_NODE_TYPES.LogicalExpression ||
                    node.type === AST_NODE_TYPES.BinaryExpression
                ) &&
                node.operator !== "??"
            ) {
                for (const child of [node.left, node.right]) {
                    if (
                        child.type === AST_NODE_TYPES.Literal &&
                        (
                            child.value === undefined ||
                            child.value === null
                        )
                    ) {
                        //Comparison against null and undefined literal is okay
                        return false;
                    }

                    if (
                        child.type === AST_NODE_TYPES.Identifier &&
                        child.name === "undefined"
                    ) {
                        //Comparison against null and undefined literal is okay
                        return false;
                    }
                }

                let hasError = false;
                if (checkBinaryNode(node.left)) {
                    hasError = true;
                }
                if (checkBinaryNode(node.right)) {
                    hasError = true;
                }
                // if this logical operator is not the root of a logical expression
                // we only check its operands and return
                if (!isRoot) {
                    return hasError;
                }
                // if this is the root of a logical expression
                // we want to check its resulting type too
                else {
                    // ...unless there already was an error, we exit so we don"t double-report
                    if (hasError) {
                        return true;
                    }
                }
            }

            const tsNode = service.esTreeNodeToTSNodeMap.get(node);
            const type = util.getConstrainedTypeAtLocation(checker, tsNode);
            let messageId: MessageId | undefined;

            const types = inspectVariantTypes(tsutils.unionTypeParts(type));

            if (types.has("bigint")) {
                messageId = "errorBuiltInOperatorUsedWithBigInt";
            }

            if (messageId != null) {
                context.report({ node, messageId });
                return true;
            }
            return false;
        }

        /** The types we care about */
        type VariantType =
            | "bigint";

        /**
         * Check union variants for the types we care about
         */
        function inspectVariantTypes(types: ts.Type[]): Set<VariantType> {
            const variantTypes = new Set<VariantType>();

            if (
                types.some(type =>
                    tsutils.isTypeFlagSet(type, ts.TypeFlags.BigInt),
                )
            ) {
                variantTypes.add("bigint");
            }

            if (
                types.some(type =>
                    tsutils.isTypeFlagSet(type, ts.TypeFlags.BigIntLike),
                )
            ) {
                variantTypes.add("bigint");
            }

            if (
                types.some(type =>
                    tsutils.isTypeFlagSet(type, ts.TypeFlags.BigIntLiteral),
                )
            ) {
                variantTypes.add("bigint");
            }

            return variantTypes;
        }
    },
});

```

And the test,
```ts
declare let x : bigint;
declare function log (arg : any) : void;

//These should error
log(typeof x == "bigint");

log(x+x);
log(x-x);
log(x*x);
log(x/x);
log(x**x);
log(x%x);
//log(+x);
log(-x);

log(++x);
log(--x);
log(x++);
log(x--);

log(x==x);
log(x!=x);
log(x===x);
log(x!==x);
log(x>x);
log(x<x);
log(x>=x);
log(x<=x);

//These should not error
log(x==undefined);
log(x!=undefined);
log(x===undefined);
log(x!==undefined);

log(undefined==x);
log(undefined!=x);
log(undefined===x);
log(undefined!==x);

log(x==null);
log(x!=null);
log(x===null);
log(x!==null);

log(null==x);
log(null!=x);
log(null===x);
log(null!==x);
```

I am using [`eslint-plugin-local`](https://www.npmjs.com/package/eslint-plugin-local) to load the custom rule.

**Comment by bradzacher** [comment]
I'm not sure I understand - what are you able to do with a bigint if you can't do any operations with it? Do you rely upon some form of a library to do any operations?

Why use the `bigint` type at all then? Why not just use some type defined by the framework?
That'd get you most of the way there and would make most of these an un-ignorable compile time error instead of an ignorable lint time one.

[repl](https://www.typescriptlang.org/play/?target=99#code/C4TwDgpgBACg9gGxAMwJYIRAJgIVQc1QDtgoBeKAbygH0aAjAJwEMisAuKIiANwkagBfANwAoLBADGCZo2jIArkUnBUcIlElzmwCHkIkAFEQUBbTidP1+UAD5R6BYsACUneEjQZs+52NGYpAAe5Jraur5GAIxELmKBUCChWhA6ek7RcaKiAPQ5ACoAFhAAztAlhXAKCFhQ-IxwjKKoyFCGoJBwrSFkFABEjgbAfS5UolATUEHCUHlQHRBdU+S9XLz8ooLZQQDU06JBALT7QQBUJzknp+diQQCk+zt7YofHt0+3r-5zwMVlULJoH0eMwEKgsH0oHBICxVEQSlBkI0AUQkgsADQOBSkJECIjqQ6WayMBG-aAghAKaAAd3QCAc0AAcsxGaJJOoSqRmOQAAXPNkc0j0UJvKDZObEKD5ADKiSqmlYAIQ1OYIAR7NMYEBUO4UApVPmcBRcDJAlpv3mf2g0P4OkaJQOvROAEIyCdem7bq7PbkcpbStaYXaSVBaRguCa6owGgIWlBMMguWwoIwCIUuXIoa0yVASsxTNAFgcAHwnAA8J2LPqCZZ9vqKAdzlWqtXxpHqjUdZCUEjQ3CwXu7bAgfewtw9PZHxDHB29k9HA+y8+nWCdYmX-dd+w32A92+HC+9+y7Jgwg9PCHHvQvg7IN+yF7XogvW7Ej6fL7XQA)

**Comment by AnyhowStep** [comment]
> Do you rely upon some form of a library to do any operations?

Yes.

> Why use the `bigint` type at all then? Why not just use some type defined by the framework?

One might be a library author and want to support both native and polyfilled bigint types. However, the native `bigint` would be preferred, of course. In such cases, one avoids using native bigint operations as much as possible, unless absolutely sure they're dealing with native bigint and not polyfilled bigint.

One might be an application developer and be using libraries that expect a bigint (native or polyfilled). The expected type would still be the TS `bigint`. But if the application is using polyfilled bigints for, say, Safari, then the application itself wouldn't want to use native bigint operations.

**Comment by AnyhowStep** [comment]
If we want to avoid problems without the lint rule, the only other alternative is remembering to cast to the correct type all the time,

```ts
//Some library

//If `BigInt` function is native, returns native bigint
//If `BigInt` function is polyfilled, returns polyfilled instance
export declare function factorial (n : bigint) : bigint;
```

```ts
//Some application
import {factorial} from "some-library";
import {polyfillBigInt, mul, PolyfilledBigInt} from "some-polyfill";

polyfillBigInt();

declare function getUserInput () : string;

const x = factorial(BigInt(getUserInput())) as PolyfilledBigInt;
const y = BigInt(getUserInput()) as PolyfilledBigInt;

//Gives compile error
console.log(x * y);

//OK
console.log(mul(x, y));
```

If you forget to cast, then you end up in situations where you can accidentally use the built-in operators without realizing it.

**Comment by AnyhowStep** [comment]
> what are you able to do with a bigint if you can't do any operations with it? 

`bigint` effectively becomes an opaque type, for the most part. It becomes more like a handle to a resource, than the actual resource itself. You can still manipulate it and get instances to it through helper functions from the polyfill library

**Comment by bbarry** [comment]
but:
```typescript
// big-int-facade.ts
export * from "some-polyfill";
 // eslint-disable-next-line no-restricted-syntax
export const produceBigInt = (input: any) => BigInt(input) as unknown as PolyfilledBigInt;

//Some application
import {factorial} from "some-library";
import {polyfillBigInt, mul, produceBigInt} from "src/facades/big-int-facade";

polyfillBigInt();

declare function getUserInput () : string;

const x = factorial(produceBigInt(getUserInput()));
const y = produceBigInt(getUserInput());

// @ts-expect-error
console.log(x * y);

//OK
console.log(mul(x, y));


// eslintrc:
```json
{
    "rules": {
        "no-restricted-syntax": [
            "error",
            {
                "selector": "CallExpression[callee.name='BigInt'][arguments.length==1]",
                "message": "use facade 'produceBigInt' instead."
            }
        ]
    }
}
```

**Comment by bradzacher** [comment]
I don't quite understand - if the library supports `bigints`, then it likely uses native operations on a `bigint`, right? So passing a polyfilled `bigint` into that function will break the library?

Unless, ofc, the lib use the same polyfill library under the hood, that is.

If it does use the polyfill library, then you can just augment the module with declarations that support your polyfill type:

```ts
declare module "some-library" {
  interface PolyfilledBigInt { __bigintBrand: never }
  export function factorial(input: PolyfilledBigInt): PolyfilledBigInt;
}

/// myfile.ts
import { factorial } from 'some-library';

const val = producePolyfilledBigint();
const result = factorial(val); // typeof result === PolyfilledBigInt
```

then you can use rules like `no-restricted-syntax` to ban `BigInt` constructor, and use our `ban-types` rule to prevent your codebase from using the `bigint` or `BigInt` types.

This way you don't have the `bigint` type anywhere in your codebase, because you're not actually using the `bigint` type.

**Comment by AnyhowStep** [comment]
> if the library supports bigints, then it likely uses native operations on a bigint, right?

Not quite. It only uses native operations if it detects that `BigInt` is native. Otherwise, it converts the object to its own internal representation. Either implemented itself or with another polyfill library. The exact mechanism doesn't matter.

You can easily do this,
```ts
function foo (x : bigint) {
  if (typeof x == "bigint") {
    return fooNative(x);
  } else {
    //BigInt is polyfilled
    return BigInt(
      fooPolyfilled(
        new SomeInternalDetail(x.toString())
      ).toString()
    );
  }
}
```

I think it is reasonable to expect `x.toString()` without a radix to return a base 10 string, whether native or polyfilled.

> So passing a polyfilled bigint into that function will break the library?

No.

> Unless, ofc, the lib use the same polyfill library under the hood, that is.

The application and library don't necessarily have to use the same polyfill library.

The code snippet shows a way for a library to support both native and polyfilled, and to return the application's polyfilled BigInt instance, even if they use different implementations.
